src/test/clone/test_fork.rs

```diff
@@ -954,7 +965,243 @@ fn test_waitid_sets_signal_dumped_info() -> anyhow::Result<()> {
     })
 }
 
+/// Core, minimal functionality for fork+exec
+fn test_fork_exec_and_reap(python_path: &Path, exit_code: i32) -> anyhow::Result<()> {
+    run_test_in_subprocess(|| {
+        let clone_res = unsafe { linux_api::sched::fork() }.unwrap();
+        let child_pid = match clone_res {
+            CloneResult::CallerIsChild => {
+                let path = CString::new(python_path.as_os_str().as_bytes()).unwrap();
+                let script = CString::new(format!(
+                    r#"
+import os
+os._exit({exit_code})
+                "#
+                ))
+                .unwrap();
+                let args = vec![path.clone(), CString::new("-c").unwrap(), script];
+                unsafe { libc::execv(path.as_ptr(), execv_argvec(&args).as_ptr()) };
+                unreachable!()
+            }
+            CloneResult::CallerIsParent(child_pid) => child_pid,
+        };
+
+        let child_pid = nix::unistd::Pid::from_raw(child_pid.as_raw_nonzero().get());
+        assert_eq!(
+            nix::sys::wait::waitpid(Some(child_pid), None).unwrap(),
+            nix::sys::wait::WaitStatus::Exited(child_pid, exit_code)
+        );
+    })
+}
+
+/// If `exec` is performed from a non-thread-group-leader, that thread becomes the
+/// new thread-group-leader. Its thread-id is updated to match the pid.
+fn test_fork_exec_from_thread_reassigns_tid(python_path: &Path) -> anyhow::Result<()> {
+    run_test_in_subprocess(|| {
+        let clone_res = unsafe { linux_api::sched::fork() }.unwrap();
+        let child_pid = match clone_res {
+            CloneResult::CallerIsChild => {
+                let path = CString::new(python_path.as_os_str().as_bytes()).unwrap();
+                std::thread::spawn(move || {
+                    let pid = unsafe { libc::getpid() };
+                    // Older libc's don't have libc::gettid, so we use the syscal wrapper instead.
+                    let tid = i32::try_from(unsafe { libc::syscall(libc::SYS_gettid) }).unwrap();
+                    assert!(pid != tid);
+                    let script = CString::new(format!(
+                        r#"
+# python doesn't expost gettid; use ctypes to invoke libc
+print("execve'd from pid:{pid} tid:{tid}")
+import ctypes
+libc = ctypes.CDLL("libc.so.6", use_errno=True)
+pid = libc.getpid()
+tid = libc.syscall({SYS_gettid})
+print(f"to pid:{{pid}} tid:{{tid}}")
+# pid should be unchanged
+assert pid == {pid}, f"Expected pid {pid}; got {{pid}}"
+# tid should now equal pid
+assert tid == {pid}, f"Expected tid {pid}; got {{tid}}""#,
```

Comment:
this is kind of hard to read; maybe give the python variables a different name than the rust code?

Reply:
Done